File: palindrome_checker/palindrome_checker.js
const accepted_characters = {}

for (let i = 48, _pj_a = 58; i < _pj_a; i += 1) {
	accepted_characters[i] = 1
}
for (let i = 97, _pj_a = 123; i < _pj_a; i += 1) {
	accepted_characters[i] = 1
}

function palindrome(s) {
	let formatted_str, alphanumeric_formatted_str
	formatted_str = s.replace(/ /g, '').toLowerCase()
	alphanumeric_formatted_str = []
	for (let i = 0; i < formatted_str.length; i++) {
		if (accepted_characters[formatted_str.charCodeAt(i)]) {
			alphanumeric_formatted_str.push(formatted_str[i])
		}
	}

	alphanumeric_formatted_str = alphanumeric_formatted_str.join()

	//   counts_map = {}
	//   flag = false

	//   for (let i = 0; i < alphanumeric_formatted_str.length; i++) {
	//     if (counts_map[alphanumeric_formatted_str[i]]) {
	//       counts_map[alphanumeric_formatted_str[i]]++
	//     } else {
	//       counts_map[alphanumeric_formatted_str[i]] = 1
	//     }
	//   }

	//   let counts = Object.values(counts_map)

	//   for (let i = 0; i < counts.length; i++) {
	//     if (counts[i] % 2 !== 0) {
	//       if (flag) {
	//         return false
	//       } else {
	//         flag = true
	//       }
	//     }
	//   }
	//   return true
	// }

	let [start, end] = [0, alphanumeric_formatted_str.length - 1]
	let mid = parseInt((end + 1) / 2)
	while (start <= mid) {
		if (alphanumeric_formatted_str[start] !== alphanumeric_formatted_str[end]) {
			return false
		}
		start += 1
		end -= 1
	}
	return true
}

console.log(palindrome('_rcar'))

File: palindrome_checker/palindrome_checker.py
accepted_characters = {}

for i in range(48, 58):
    accepted_characters[i] = 1

for i in range(97, 123):
    accepted_characters[i] = 1


def palindrome(s):
    formatted_str = s.replace(" ", "").lower()
    alphanumeric_formatted_str = []
    for ch in formatted_str:
        if ord(ch) in accepted_characters:
            alphanumeric_formatted_str.append(ch)

    alphanumeric_formatted_str = "".join(alphanumeric_formatted_str)

    # counts_map = Counter(alphanumeric_formatted_str)
    # flag = False
    # for v in counts_map.values():
    #     if v % 2 != 0:
    #         if flag:
    #             return False
    #         else:
    #             flag = True

    # return True

    start, end = 0, len(alphanumeric_formatted_str) - 1
    mid = (end+1)//2
    while start <= mid:
        if alphanumeric_formatted_str[start] != alphanumeric_formatted_str[end]:
            return False
        start += 1
        end -= 1
    return True

print(palindrome("_racar"))
